### PR TITLE
order TestWatcher to be outermost

### PR DESCRIPTION
**Pull request by panchenko**
As a possible way to address ordering in #1436

The change of spaces is done to make `LoggingTestRule` consistent with `LoggingTestWatcher`.

### CODE REVIEW

**Comment by kcooney** [comment]
Interesting idea. I worry that this could break existing tests. If it does, there may not be a way for the user to fix their tests.

**Comment by panchenko** [comment]
I agree it's tricky. Also agree with deprecating `ExpectedException`.

A possible improvement could be doing this only if `@Rule` has default order.

**Comment by kcooney** [comment]
@panchenko so perhaps a `DefaultOrder` annotation?

**Comment by panchenko** [comment]
There are 2 parts: 
* preferred/default order for some rules
* how it can be overridden if needed, e.g. if `@Rule` has `order != Rule.DEFAULT_ORDER`

**Comment by kcooney** [comment]
Agreed. I am just suggesting that the class level annotation have an attribute which is the default order value.

**Comment by stefanbirkner** [comment]
TLDR: I propose to not merge this pull request and use `@Rule(order = Integer.MIN_VALUE)` for solving #1436. See pull request #1501.

I think that the annotation `@OutermostRule` adds a lot of complexity considering the problem that it solves. Currently we have one mechanism for ordering rules: the `order` attribute of the `@Rule` annotation. (You could also consider a ordering mechanism that rules created by methods are collected first, but for now I leave this aside.) `@OutermostRule` adds a second mechanism by separating the rules into two classes and enforcing the following order: outermost - low order, outermost - high order, standard rule - low order, standard rule - high order. Unfortunately this is not apparent to the reader of a test:

    public class SomeTest {
      @Rule(order = 1)
      public final Testrule firstRule = new FirstRule();

      @Rule(order = 2)
      public final Testrule secondRule = new SecondRule(); //this is a outermost rule

      @Rule(order = 3)
      public final Testrule thirdRule = new ThirdRule();

      @Rule(order = 4)
      public final Testrule fourthRule = new FourthRule(); //this is a outermost rule

      ...
    }

The order of these rules is `secondRule`, `fourthRule`, `firstRule`, `thirdRule`. In order to understand this the reader must know about `@OutermostRule` and they must know that `SecondRule` and `ThirdRule` are annotated with `@OutermostRule`. The same rule order can be achieved by 

    public class SomeTest {
      @Rule(order = 1)
      public final Testrule firstRule = new FirstRule();

      @Rule(order = Integer.MAX_VALUE)
      public final Testrule secondRule = new SecondRule();

      @Rule(order = 3)
      public final Testrule thirdRule = new ThirdRule();

      @Rule(order = Integer.MAX_VALUE + 1)
      public final Testrule fourthRule = new FourthRule();

      ...
    }

The reader must only know about rule ordering. What is lost is that without the `order` attribute the rules `secondRule` and `fourthRule` would still enclose `firstRule` and `thirdRule`.

The reason for implementing this annotation was the problem that TestWatcher should always be the outermost rule (see #1436). Currently I don't know another use case for `@OutermostRule`. But the `TestWatcher` problem can already be solved by using `@Rule(order = Integer.MIN_VALUE)` without implementing a second mechanism for ordering.

Unfortunately the `order` approach has a drawback, too. The test writer is now responsible for adding the `order` attribute to the rule. The best we can do is documentation in the class' Javadoc. Hopefully the popular IDEs implement a warning when a `TestWatcher` is used without an explicit `order` attribute.

If we still want `TestWatcher`s automatically to be the outermost rules than we should do this without the additional annotation. We can check rules for being a `TestWatcher` and adjust the order of rules accordingly.

**Comment by kcooney** [comment]
@stefanbirkner my proposal was to have the `order` attribute of `Rule` override this behavior of having`TestWatcher` be the outermost rule.

That being said, I agree that adding documentation might be the best approach.

**Comment by panchenko** [comment]
I've changed annotation to `@DefaultOrder`.
The next question if it should be specified for other rules, like `Verifier`, etc.

**Comment by stefanbirkner** [comment]
The changed code shows pretty clear how `@DefaultOrder` works. Thanks for that piece of work. Nevertheless I'm still in favor for fixing the problem with documentation and hopefully IDE support.

Although the classification issue is gone it has been replaced by a new issue: the value of `order` is hard to choose because it is not absolute. Only for the outermost and the innermost rule the `order` is kind of clear.

The rest of my preceding comment still applies to `@DefaultOrder`.

**Comment by kcooney** [comment]
If we solve this problem with code, the current pull is the way I would like to solve it.

The value for `DefaultOrder.order` is hard to choose, but that doesn't bother me because it's a default value which users can override via `Rule.order`

What bothers me is we would yet again be changing the order that things happen for existing tests. For large code bases these kids of changes slow down people from upgrading.

**Comment by stefanbirkner** [comment]
The change of the order of existing tests together with the added complexity makes me reluctant to merge this pull request. With the doc fix all existing tests keep successful while new tests can fix the `TestWatcher` problem and even the `ErrorCollector/ExpectedException` problem by setting `order`. If we go for the doc fix then I create a doc update for `ErrorCollector/ExpectedException`, too.

**Comment by panchenko** [comment]
I think an update for those classes documentation would be useful in any case (with or without this).

**Comment by kcooney** [comment]
Since #1501 merged, and people were at best luke-warm on this approach, I am closing.
